node/impl/full/trace.go

```diff
@@ -0,0 +1,515 @@
+	ts, err := e.getTipsetByBlockNr(ctx, blkNum, false)
+	if err != nil {
+		return nil, err
+	}
+
+	_, trace, err := e.StateManager.ExecutionTrace(ctx, ts)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to compute base state: %w", err)
+	}
+
+	tsParent, err := e.ChainAPI.ChainGetTipSetByHeight(ctx, ts.Height()+1, e.Chain.GetHeaviestTipSet().Key())
+	if err != nil {
+		return nil, fmt.Errorf("cannot get tipset at height: %v", ts.Height()+1)
+	}
+
+	msgs, err := e.ChainGetParentMessages(ctx, tsParent.Blocks()[0].Cid())
+	if err != nil {
+		return nil, err
+	}
+
+	cid, err := ts.Key().Cid()
+	if err != nil {
+		return nil, err
+	}
+
+	blkHash, err := ethtypes.EthHashFromCid(cid)
+	if err != nil {
+		return nil, err
+	}
+
+	allTraces := make([]*TraceBlock, 0, len(trace))
+	for _, ir := range trace {
+		// ignore messages from f00
+		if ir.Msg.From.String() == "f00" {
+			continue
+		}
+
+		idx := -1
+		for msgIdx, msg := range msgs {
+			if ir.Msg.From == msg.Message.From {
+				idx = msgIdx
+				break
+			}
+		}
+		if idx == -1 {
+			log.Warnf("cannot resolve message index for cid: %s", ir.MsgCid)
+			continue
+		}
+
+		txHash, err := e.EthGetTransactionHashByCid(ctx, ir.MsgCid)
+		if err != nil {
+			return nil, err
+		}
+		if txHash == nil {
+			log.Warnf("cannot find transaction hash for cid %s", ir.MsgCid)
+			continue
+		}
+
+		traces := []*Trace{}
+		err = buildTraces(&traces, nil, []int{}, ir.ExecutionTrace, int64(ts.Height()))
+		if err != nil {
+			return nil, xerrors.Errorf("failed when building traces: %w", err)
+		}
+
+		traceBlocks := make([]*TraceBlock, 0, len(trace))
+		for _, trace := range traces {
+			traceBlocks = append(traceBlocks, &TraceBlock{
+				Trace:               trace,
+				BlockHash:           blkHash,
+				BlockNumber:         int64(ts.Height()),
+				TransactionHash:     *txHash,
+				TransactionPosition: idx,
+			})
+		}
+
+		allTraces = append(allTraces, traceBlocks...)
+	}
+
+	return allTraces, nil
+}
+
+func (e *EthTrace) TraceReplayBlockTransactions(ctx context.Context, blkNum string, traceTypes []string) (interface{}, error) {
+	if len(traceTypes) != 1 || traceTypes[0] != "trace" {
+		return nil, fmt.Errorf("only 'trace' is supported")
+	}
+
+	ts, err := e.getTipsetByBlockNr(ctx, blkNum, false)
+	if err != nil {
+		return nil, err
+	}
+
+	_, trace, err := e.StateManager.ExecutionTrace(ctx, ts)
+	if err != nil {
+		return nil, xerrors.Errorf("failed when calling ExecutionTrace: %w", err)
+	}
+
+	allTraces := make([]*TraceReplayBlockTransaction, 0, len(trace))
+	for _, ir := range trace {
+		// ignore messages from f00
+		if ir.Msg.From.String() == "f00" {
+			continue
+		}
+
+		txHash, err := e.EthGetTransactionHashByCid(ctx, ir.MsgCid)
+		if err != nil {
+			return nil, err
+		}
+		if txHash == nil {
+			log.Warnf("cannot find transaction hash for cid %s", ir.MsgCid)
+			continue
+		}
+
+		t := TraceReplayBlockTransaction{
+			Output:          hex.EncodeToString(ir.MsgRct.Return),
```

Comment:
We need to decode this return value, unfortunately:

1. If empty, treat it as empty bytes.
2. If non-empty, decode it as a cbor byte array, then hex-encode it.

Take a look at:

https://github.com/filecoin-project/lotus/blob/cbbb76b2f095782e5c4d717b5c2603ccd41d21ff/node/impl/full/eth.go#L695-L698

Reply:
Well.... ok, it's worse:

1. We need to do this _if_ the method number is "InvokeEVM" (InvokeContract).
2. If we're invoking one of the "create" methods on the EAM.

Otherwise... we need to do the whole "make it look like solidity ABI" dance.

E.g., take a look at

https://github.com/filecoin-project/lotus/blob/cbbb76b2f095782e5c4d717b5c2603ccd41d21ff/node/impl/full/eth.go#L2108

https://github.com/filecoin-project/lotus/blob/cbbb76b2f095782e5c4d717b5c2603ccd41d21ff/chain/types/ethtypes/eth_transactions.go#L67

Reply:
> Otherwise... we need to do the whole "make it look like solidity ABI" dance.

@Stebalien could to clarify what you mean by that, its a bit unclear how the (`newEthTxFromSignedMessage`,`EthTxFromSignedEthMessage`) would apply in this case

Reply:
I mean, take a look at how those functions interpret message parameters and return values. I.e.:

- We decode parameters and return values as "cbor bytes" (usually).
- We treat messages to the CreateExternal (4) method on the EAM specially. They are contract creations, not normal messages. I assume that the EVM traces handle this specially as well.

Reply:
@Stebalien for the ir.MsgRct.Return there is no codec included in the receipt and when I use UnmarshalCBOR something like this:

```go
var output abi.CborBytes
if len(ir.MsgRct.Return) > 0 && ir.MsgRct.ExitCode.IsSuccess() {
    if err := output.UnmarshalCBOR(bytes.NewReader(ir.MsgRct.Return)); err != nil {
	log.Warnf("failed to unmarshal cbor return value '%s': %w", string(ir.MsgRct.Return), err)
	//return nil, xerrors.Errorf("failed to unmarshal cbor return value '%s': %w", string(ir.MsgRct.Return), err)
    }
    // pad with zero bytes if smaller than 32 bytes
    output = append(make([]byte, 32-len(output), 32), output...)
}
```

I get an error `expected byte array`, similarly it does not work to use `cbg.ReadByteArray` either

Reply:
For the codec: So, I think you can use the top-level "return" in the trace. That should be equivalent to the receipt, and should actually include the codec.

However, I _think_ the issue here is that you're decoding non-EVM return values? I.e., it's a return value from a call on a method that's not `InvokeEVM`, but some other thing. In that case, instead of _decoding_ it as cbor-bytes, we need to _encode_ it as solidity-ABI.

Or am I missing something.

Reply:
Ah, using the top level return in the trace works